### PR TITLE
Is/use pipeline status (backend)

### LINKED ISSUE #1781: Get task status from Celery to allow run button status updates
Task status can be retrieved from Celery using AsyncResult.
Each pipeline creates a series of tasks. the task without a parent is the main task.

A new endpoint shall be created so the frontend can update the status of the run button. (which would also prevent from having unwanted multiple runs)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

replaces #1807 

fixes #1781

**Webserver API changes:**
- adds running state of pipeline in /projects/{uuid} endpoint with possible values as **NOT_STARTED, PUBLISHED, PENDING, STARTED, RETRY, SUCCESS, FAILURE** (updated project-v0.0.1.json schema)
- individual node state is saved using the same enumeration and can be retrieved in the **workbench** field (updated project-v0.0.1.json schema)
- computation/{project_uuid} endpoint removed (unused)
- computation/{project_uuid}/start now returns **403-Forbidden** if a pipeline is in PUBLISHED, PENDING, STARTED, RETRY state

**DB changes**
-  comp_tasks and comp_pipeline now uses a postgresql type StateType instead of an integer for easier debugging (migration shall be handled by migration scripts)
- comp_tasks trigger is now also checking when a node state is changed

**Webserver behavior changes**
- validation includes state of project/nodes
- update node state from when database triggers
- when computation pipeline is started, comp_tasks entries state is set to PUBLISHED with a **60 seconds** timeout (e.g. during that time the pipeline **cannot** be restarted)
- publication timeout in celery is a settings modifiable through ENVs
- when the state of a node/pipeline changes, a websocket event is fired to allow frontend client(s) to update their state (@odeimaiz to implement)

**Sidecar behavior changes**
- only start a task if in PENDING or FAILED state
- initial task sets other tasks from PUBLISHED to PENDING (e.g. acknowledges new pipeline is received)

Bonus:
- fixed requirements in sidecar (alembic, pytest-icdiff)

## Checklist

- [x] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: services/web/server/src/simcore_service_webserver/computation_api.py [hunk 1]
```diff
@@ -437,6 +445,135 @@ async def update_pipeline_db(
...
+    config = app[APP_CONFIG_KEY][CONFIG_RABBIT_SECTION]
+    rabbit = RabbitConfig(**config)
+    return rabbit.publication_timeout
+
+
+async def _set_tasks_in_tasks_db_as_published(db_engine: Engine, project_id: str):
+    query = (
+        # pylint: disable=no-value-for-parameter
+        comp_tasks.update()
+        .where(and_(comp_tasks.c.project_id == project_id))
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_api.py:469]: I think you forgot this ``and_``

File: services/web/server/src/simcore_service_webserver/computation_api.py [hunk 2]
```diff
@@ -437,6 +445,135 @@ async def update_pipeline_db(
...
+        "Task (task=%s, user_id=%s, project_id=%s) submitted for execution.",
+        task.task_id,
+        user_id,
+        project_id,
+    )
+    return task.task_id
+
+
+def _from_celery_state(celery_state) -> RunningState:
+    CELERY_TO_RUNNING_STATE = {
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_api.py:505]: MINOR: put these maps outside so they are init only once. Same with ``DB_TO_RUNNING_STATE``

File: services/web/server/src/simcore_service_webserver/computation_api.py [hunk 3]
```diff
@@ -437,6 +445,135 @@ async def update_pipeline_db(
...
+        set_states = {state[0] for state in task_states.values()}
+        last_update = next(
+            iter(sorted([time[1] for time in task_states.values()], reverse=True))
+        )
+        if RunningState.published in set_states:
+            if (now - last_update).seconds > get_celery_publication_timeout(app):
+                return RunningState.not_started
+        if len(set_states) == 1:
+            # this is typically for success, pending, published
+            return next(iter(set_states))
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_api.py:564]: MINOR: `` return set_states.pop()``  ??

File: services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py [hunk 4]
```diff
@@ -2,29 +2,95 @@
...
+                # value output
+                return True
+    return False
+
+
+def _is_state_changed(current_state: str, new_state: str) -> bool:
+    return current_state != new_state
+
+
+from .computation_api import convert_state_from_db
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py:48]: MINOR: all imports on top

File: services/web/server/tests/unit/with_dbs/fast/test_computation_api.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        computation_api,
+        "get_celery_publication_timeout",
+        return_celery_publication_timeout,
+    )
+
+
+fake = faker.Faker()
+
+
+@pytest.mark.parametrize(
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/fast/test_computation_api.py:64]: very nice!

File: services/web/server/tests/unit/with_dbs/fast/test_computation_api.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        return 60
+
+    monkeypatch.setattr(
+        computation_api,
+        "get_celery_publication_timeout",
+        return_celery_publication_timeout,
+    )
+
+
+fake = faker.Faker()
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/fast/test_computation_api.py:61]: MINOR: put on top

**Review by pcrespov — COMMENTED** [review]
Will check more tomorrow...

File: packages/models-library/src/models_library/projects.py [hunk 7]
```diff
@@ -211,29 +251,24 @@ class Project(BaseModel):
...
+    "ProjectRunningState",
     "ProjectLocked",
+    "RunningState",
     "Owner",
 ]
+
+
+if __name__ == "__main__":
+
+    with open(current_file.with_suffix(".json"), "wt") as fh:
```
Comment (GitHK) [inline packages/models-library/src/models_library/projects.py:273]: At this point move the `current_file` definition in this block

File: packages/models-library/src/models_library/projects.py [hunk 8]
```diff
@@ -12,6 +12,17 @@
 DATE_RE = r"\d{4}-(12|11|10|0?[1-9])-(31|30|[0-2]?\d)T(2[0-3]|1\d|0?[0-9])(:(\d|[0-5]\d)){2}(\.\d{3})?Z"
 
 
+class RunningState(str, Enum):
+    unknown = "UNKNOWN"
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/projects.py:16]: a bad habit, but ok I'll comply

File: services/web/server/src/simcore_service_webserver/computation_api.py (hunk 3 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/computation_api.py:564]: why would I do that?

File: packages/models-library/src/models_library/projects.py [hunk 9]
```diff
@@ -211,29 +251,24 @@ class Project(BaseModel):
...
 __all__ = [
     "ProjectState",
+    "ProjectRunningState",
     "ProjectLocked",
+    "RunningState",
     "Owner",
 ]
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/projects.py:271]: I'll remove it.

File: packages/models-library/src/models_library/projects.py [hunk 10]
```diff
@@ -161,6 +178,29 @@ class Config:
...
+
+
+class ProjectLocked(BaseModel):
+    value: bool = Field(
+        ..., description="True if the project is locked by another user"
+    )
+    owner: Optional[Owner] = Field(None, description="The user that owns the lock")
+
+
+class ProjectRunningState(BaseModel):
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/projects.py:193]: well I use the value field by design. so if I have to add something later I already have an object.

[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/projects.py:193]: but that is a good point still.

File: services/sidecar/src/simcore_service_sidecar/core.py [hunk 11]
```diff
@@ -199,13 +226,18 @@ async def inspect(
...
         await executor.run()
         next_task_nodes = list(graph.successors(node_id))
-        run_result = SUCCESS
+        run_result = StateType.SUCCESS
     except asyncio.CancelledError:
         log.warning("Task has been cancelled")
         raise
 
     finally:
+        await rabbit_mq.post_log_message(
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/sidecar/src/simcore_service_sidecar/core.py:235]: it's ok celery takes care of the exceptions.

**Review by GitHK — COMMENTED** [review]
Running it, behaves much better than before. I can abuse the run button and still have some results. So it looks ok. I've seen some other things which are not reported but they are trivial.

File: packages/models-library/src/models_library/projects.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/projects.py]: it's removed

File: services/web/server/src/simcore_service_webserver/computation_api.py (hunk 3 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/computation_api.py:568]: no strong reason .. i just find it more readable to get the last element of your set. It is a MINOR comment

**Review by pcrespov — APPROVED** [review]
Release the beast! :-)

**Review by GitHK — APPROVED** [review]

**Comment by ignapas** [comment]
Frontend WiP

**Pull request merged**